integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -718,12 +734,14 @@ def compare_backends(self, trace_function: Callable[[TracedModule], None],
 
     # Compare each target trace of trace_function with the reference trace.
     failed_backend_indices = []
+    error_messages = []
     for i, tar_trace in enumerate(tar_traces):
       logging.info("Comparing the reference backend '%s' with '%s'",
                    ref_trace.backend_id, tar_trace.backend_id)
-      traces_match = Trace.compare_traces(ref_trace, tar_trace)
+      traces_match, errors = Trace.compare_traces(ref_trace, tar_trace)
       if not traces_match:
         failed_backend_indices.append(i)
+        error_messages.extend(errors)
```

Comment:
These are now going to be printed without the context of which backend they came from. I think that should probably be included. Related to that, it would be good to propagate the failures up in a more structured way (not strings) so they can be displayed in this failure rather than in the logs.

Reply:
**Re: Organizing by backend**
It wouldn't be too difficult to organize the errors by the target backend they came from – I just didn't include it because we never use multiple target backends in our tests, and when someone runs a `_manual` target the logs are always printed directly into the console, so they don't rely on the `AssertionError` alone to understand the failure.

**Re: All logs in the `AssertionError`**
I don't think that it would be useful to show all of the information we log about a failure in the `AssertionError`. Instead I think it would be nice to have a high-level overview of what the failure was, which someone use to decide if they need to get more information from the logs. Also, for most of our tests the full logs about a failure exceeds 300 lines, and are only useful for more in-depth debugging.

We can infer the trace function/method that is being compared already from what the metadata that the `AssertionError` prints. I updated the example message above to use a real failure instead of one that I made up to show this.